src/PluginHandler.cpp

```diff
@@ -0,0 +1,1065 @@
+    string suffix = path.substr(slashpos + 1);
+    if (installPaths.find(location) == installPaths.end()
+        && archive_entry_filetype(entry) == AE_IFREG
+    ) {
+        location = "unknown";
+    }
+    string dest = installPaths[location] + "/" + suffix;
+    archive_entry_set_pathname(entry, dest.c_str());
+}
+
+
+
+static void linux_entry_set_install_path(struct archive_entry* entry,
+                                         pathmap_t installPaths)
+{
+    using namespace std;
+
+    string path = archive_entry_pathname(entry);
+    int slashes = count(path.begin(), path.end(), '/');
+    if (slashes < 3) {
+        archive_entry_set_pathname(entry, "");
+        return;
+    }
+    int slashpos = path.find_first_of('/', 1);
+    string prefix = path.substr(0, slashpos);
+    path = path.substr(prefix.size() + 1);
+    if (startswith(path, "usr/")) {
+        path = path.substr(strlen("usr/"));
+    }
+    if (startswith(path, "local/")) {
+        path = path.substr(strlen("local/"));
+    }
+    slashpos = path.find_first_of('/');
+    string location = path.substr(0, slashpos);
+    string suffix = path.substr(slashpos + 1);
+    if (installPaths.find(location) == installPaths.end()
+        && archive_entry_filetype(entry) == AE_IFREG
+    ){
+        location = "unknown";
+    }
+    string dest = installPaths[location] + "/" + suffix;
+    archive_entry_set_pathname(entry, dest.c_str());
+}
+
+
+static void apple_entry_set_install_path(struct archive_entry* entry,
+                                         pathmap_t installPaths)
+{
+    using namespace std;
+
+    const string base = PluginPaths::getInstance()->Homedir()
+        + "/Library/Application Support/OpenCPN";
+
+    const string path = archive_entry_pathname(entry);
+    string dest("");
+    size_t slashes = count(path.begin(), path.end(), '/');
+    if (slashes < 3) {
+        archive_entry_set_pathname(entry, "");
+        return;
+    }
+    auto parts = split(path, "Contents/Resources");
+    if (parts.size() >= 2) {
+        dest = base  + "/Contents/Resources" + parts[1];
+    }
+    if (dest == "") {
+        parts = split(path, "Contents/SharedSupport");
+        if (parts.size() >= 2) {
+            dest = base + "/Contents/SharedSupport"  + parts[1];
+        }
+    }
+    if (dest == "") {
+        parts = split(path, "Contents/PlugIns");
+        if (parts.size() >= 2) {
+            dest = base +  "/Contents/PlugIns" + parts[1];
+        }
+    }
+    if (dest == "" && archive_entry_filetype(entry) == AE_IFREG) {
+        // Drop uninstalled directories.
+        dest = installPaths["unknown"] + "/" + path;
+    }
+    archive_entry_set_pathname(entry, dest.c_str());
+}
+
+
+
+static void entry_set_install_path(struct archive_entry* entry,
+                                   pathmap_t installPaths)
+{
+    const auto osSystemId = wxPlatformInfo::Get().GetOperatingSystemId();
+    const std::string src = archive_entry_pathname(entry);
```

Comment:
What happens when the path is e.g. `/foobar` (absolute) or `foo/../../../../../../../foobar`?

Reply:
Good question for sure.

As for `/foobar` the paths comes from a tarball, and such paths are not absolute. 

In the more general case, the tarballs are basically created by `make install` and are supposed to be sane. That is, paths cannot really be anything. Which, in turn, boils down to missing documentation for the plugins: metadata semantics, icons, tarball layout.

For now, assuming that the tarball paths are as created by make install, different on different platforms. To make things more interesting this is not really used and tested today besides linux, and there is at least one wide-spread, copy-pasted  bug on MacOS

Reply:
BTW: There is now a plugins/README.md with some info on tarball layout